src/repository/user.local.pg.repository.ts

```diff
@@ -0,0 +1,151 @@
+import Database from './database.interface';
+import UserRemote from '../model/entity/user.remote.type';
+import UserLocal from '../model/entity/user.local.type';
+import ListUser from '../model/request/list.user.type';
+
+const { PreparedStatement } = require('pg-promise');
+
+export const getUserByName = (database: Database) => async (username: string): Promise<UserLocal> => database.handler.oneOrNone(new PreparedStatement(
+  {
+    name: 'find-user-by-name',
+    text: 'SELECT * FROM users WHERE name = ($1)',
+    values: [username],
+  },
+));
+
+export const getUsers = (database: Database) => async (filters: ListUser): Promise<UserRemote[]> => {
+  let query = 'SELECT DISTINCT '
+        + 'users.name, '
+        + 'locations.name as location, '
+        + 'array_agg(languages.name) as languages '
+        + 'FROM users '
+        + 'LEFT JOIN locations '
+        + 'ON users.location_id = locations.id '
+        + 'LEFT JOIN user_languages '
+        + 'ON users.id = user_languages.user_id '
+        + 'LEFT JOIN languages '
+        + 'ON user_languages.language_id = languages.id '
+        + 'WHERE true '; // chain filters 0, 1 and 1+
+
+  if (filters.location) {
+    query += database.configs.as.format(
+      'AND locations.name = $/location/ ',
+      { location: filters.location },
+    );
+  }
+
+  if (filters.languages) {
+    query += database.configs.as.format(
+      'AND languages.name IN ($/languages:csv/) ',
+      { languages: filters.languages },
+    );
+  }
+
+  query += database.configs.as.format(
+    'GROUP BY users.id, locations.name LIMIT $/limit/ OFFSET $/offset/ ',
+    {
+      limit: filters.limit,
+      offset: filters.offset,
+    },
+  );
+
+  return database.handler.any(new PreparedStatement(
+    {
+      name: 'list-all-users-with-filters',
+      text: query,
+    },
+  ));
+};
+
+const persistUserLanguages = async (database: Database, userId: number, languagesIds: number[]): Promise<void> => {
+  const columnSet = new database.configs.helpers.ColumnSet(
+    ['language_id', 'user_id'],
+    { table: 'user_languages' },
+  );
+
+  const values = languagesIds.map(
+    (element) => ({ language_id: element, user_id: userId }),
+  );
+
+  const query = database.configs.helpers.insert(values, columnSet)
+    + ' ON CONFLICT DO NOTHING';
+
+  return database.handler.none(new PreparedStatement({
+    name: 'persist-user-languages',
+    text: query,
+  }));
+};
+
```

Comment:
Perhaps it's me, but I would not expect a function named `get` to be doing `INSERT`ions implicitly - I would either rename this function to be more clear of what it's doing, or change the code

Reply:
I don't disagree with this statement, however, for encapsulation purposes, I would like to avoid exposing the database internals to the service layer.  Naming this method ```getLocationIdAndInsertIfDoesntExists()``` is a bit obnoxious. I guess ```obtainLocationId()``` would be a good middle ground solution